### PR TITLE
bpo-38266: Revert bpo-37878: Make PyThreadState_DeleteCurrent() Internal

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
Reverting the removal of  ``PyThreadState_DeleteCurrent()`` with documentation.

https://bugs.python.org/issue38266

### CODE REVIEW

File: Doc/c-api/init.rst
Comment (vstinner) [inline Doc/c-api/init.rst:1043]: If we start to "expose" this function, would it be possible to be less strict and call PyThreadState_Clear() in PyThreadState_DeleteCurrent()? It sems like calling PyThreadState_Clear() twice is safe.

File: Include/internal/pycore_pylifecycle.h
Comment (vstinner) [inline Include/internal/pycore_pylifecycle.h]: I don't understand why do you remove this definition: it's still used, no?

File: Include/cpython/pystate.h
Comment (vstinner) [inline Include/cpython/pystate.h]: The prototype must be:

```suggestion
PyAPI_FUNC(void) PyThreadState_DeleteCurrent(void);
```

File: Modules/_tracemalloc.c
Comment (vstinner) [inline Modules/_tracemalloc.c]: I prefer to keep _PyThreadState_DeleteCurrent() here. PyThreadState_DeleteCurrent() is only a public helper function, _PyThreadState_DeleteCurrent() is the real implementation.

File: Modules/_tracemalloc.c
Comment (vstinner) [inline Modules/_tracemalloc.c]: ditto: keep _PyThreadState_DeleteCurrent()

File: Python/pystate.c
Comment (vstinner) [inline Python/pystate.c]: Please add void to prevent accepting any argument:

```suggestion
PyThreadState_DeleteCurrent(void)
```

**Comment by vstinner** [comment]
Once this change lands, it would be interested to add an unit test, but I'm not sure how to write such test. C code in _testcapi I guess.

File: Include/internal/pycore_pylifecycle.h
Comment (nanjekyejoannah) [reply Include/internal/pycore_pylifecycle.h]: I am now exposing PyThreadState_DeleteCurrent() instead of the internal  _PyThreadState_DeleteCurrent() I thought. what are you thinking?

**Comment by nanjekyejoannah** [comment]
@vstinner , I have addressed the changes. PTAL

File: Doc/c-api/init.rst
Comment (vstinner) [reply Doc/c-api/init.rst:1043]: I marked https://bugs.python.org/issue38266 as a release blocker, so the PyThreadState_Clear() question can definitely wait after 3.8.0: we can even revisit this in Python 3.9. Let's stick to adding back the function unmodified.

File: Include/internal/pycore_pylifecycle.h
Comment (vstinner) [reply Include/internal/pycore_pylifecycle.h]: _PyThreadState_DeleteCurrent() has to stay, it's part of https://pythoncapi.readthedocs.io/runtime.html effort to avoid global variables.

**Review by vstinner — APPROVED** [review]
LGTM.

File: Include/internal/pycore_pylifecycle.h
Comment (vstinner) [inline Include/internal/pycore_pylifecycle.h]: nitpick: it seems like you removed a newline or something else. you may revert that.

**Comment by vstinner** [comment]
According to https://bugs.python.org/issue38266 discussion, it's a Python 3.8 regression and so this change must be backported to 3.8. I suggest to merge this change into 3.8 and master as soon as possible, to attempt to get it into 3.8.0 final. (But it's ok if we miss the final, it can wait for 3.8.1 otherwise.)

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Usually, we let core developers merge their own pull requests. But. Joannah explicitly asked me to merge her PR if I want. Since this change fix a Python 3.8 regression and the next 3.8.0 release (which may be a rc2 or the final) is in 3 days, I prefer to merge it ASAP.

**Pull request merged**
